### ISSUE TITLE
Stop Carp from polluting the caller's namespace

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12966]
Migrated from [rt.perl.org#117989](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=117989) (status was 'resolved')

Searchable as RT117989$

**Comment by p5pRT** [comment]
### From @Hugmeir

#### Created by @Hugmeir



In perl 5\.6\, this&#8203;:

perl \-MCarp \-le 'package Foo; eval \{ Carp&#8203;::croak\(1\) \}; print keys %Foo&#8203;::;'

Used to output 'ISA'\. In 5\.8 and newer\, it outputs "ISA" and "CARP\_NOT"\.
This is because Carp&#8203;::trusts\_directly\(\) simply tries using both
of those variables without checking if they exists first\, which
ends up creating them\.
The first of the attached patches fixes that so that no new globs are
introduced to the caller's namespace\, by checking if the variable
exists in the stash before using them\.

The second patch fixes a related issue&#8203;:

&nbsp;    use Carp;
&nbsp;    package Foo;
&nbsp;    $CARP\_NOT = 1;
&nbsp;    warn \*\{$Foo&#8203;::\{CARP\_NOT\}\}\{ARRAY\} || 'undefined';
&nbsp;    eval \{ Carp&#8203;::croak\(1\) \};
&nbsp;    warn \*\{$Foo&#8203;::\{CARP\_NOT\}\}\{ARRAY\} || 'undefined';

Both should display "undefined"\, but instead the second warn
shows that the array slot got vivified by Carp\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low
    module=Carp

Site configuration information for perl 5.16.2:

Configured by hugmeir at Mon Feb 11 17:35:33 ART 2013.

Summary of my perl5 (revision 5 version 16 subversion 2) configuration:

  Platform:
    osname=darwin, osvers=12.2.1, archname=darwin-thread-multi-2level
    uname='darwin airtiger 12.2.1 darwin kernel version 12.2.1: thu oct 18
12:13:47 pdt 2012; root:xnu-2050.20.9~1release_x86_64 x86_64 '
    config_args='-de
-Dprefix=/Users/hugmeir/perl5/perlbrew/perls/perl-5.16.2 -DDEBUGGING
-Dusethreads -Uversiononly -Accflags=-Wall -Wextra
-Aeval:scriptdir=/Users/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-common -DPERL_DARWIN -Wall -Wextra -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector',
    optimize='-O3 -g',
    cppflags='-fno-common -DPERL_DARWIN -Wall -Wextra -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector'
    ccversion='', gccversion='4.2.1 Compatible Apple LLVM 4.2
(clang-425.0.24)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector'
    libpth=/usr/lib
    libs=-ldbm -ldl -lm -lutil -lc
    perllibs=-ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup
-fstack-protector'

Locally applied patches:



@INC for perl 5.16.2:
    /Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
/lib/perl5/darwin-thread-multi-2level
    /Users/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5

/Users/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/site_perl/5.16.2/darwin-thread-multi-2level
    /Users/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/site_perl/5.16.2

/Users/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/5.16.2/darwin-thread-multi-2level
    /Users/hugmeir/perl5/perlbrew/perls/perl-5.16.2/lib/5.16.2
    .


Environment for perl 5.16.2:
    DYLD_LIBRARY_PATH=/usr/local/mysql/lib:
    HOME=/Users/hugmeir
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
/bin:/Users/hugmeir/perl5/perlbrew/bin:/Users/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/usr/local/git/bin:/usr/local/mysql/bin
    PERL5LIB=/Users/hugmeir/.perlbrew/libs/perl-5.16.2@all/lib/perl5
    PERLBREW_BASHRC_VERSION=0.58
    PERLBREW_HOME=/Users/hugmeir/.perlbrew
    PERLBREW_LIB=all
    PERLBREW_MANPATH=/Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
/man:/Users/hugmeir/perl5/perlbrew/perls/perl-5.16.2/man
    PERLBREW_PATH=/Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
/bin:/Users/hugmeir/perl5/perlbrew/bin:/Users/hugmeir/perl5/perlbrew/perls/perl-5.16.2/bin
    PERLBREW_PERL=perl-5.16.2
    PERLBREW_ROOT=/Users/hugmeir/perl5/perlbrew
    PERLBREW_VERSION=0.58
    PERL_BADLANG (unset)
    PERL_LOCAL_LIB_ROOT=/Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MB_OPT=--install_base /Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
    PERL_MM_OPT=INSTALL_BASE=/Users/hugmeir/.perlbrew/libs/perl-5.16.2@all
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @Hugmeir

<details><summary>0001-Carp-Stop-polluting-the-caller-s-namespace-if-they-l.patch</summary>

```diff
From be7d89c3ced94ca13a57dff3df46566507d83617 Mon Sep 17 00:00:00 2001
From: Brian Fraser <fraserbn@gmail.com>
Date: Mon, 13 May 2013 15:57:18 -0300
Subject: [PATCH 1/2] Carp: Stop polluting the caller's namespace if they lack
 @CARP_NOT or @ISA

Turns out that this:

perl -MCarp -le 'package Foo; eval { Carp::croak(1) }; print keys %Foo::;'

Would leave two new symbols in Foo, CARP_NOT and ISA. This commit
changes trusts_directly() to check if the symbols exists in the
stash before using them.
---
 dist/Carp/lib/Carp.pm | 15 ++++++++++-----
 dist/Carp/t/Carp.t    | 12 +++++++++++-
 2 files changed, 21 insertions(+), 6 deletions(-)

diff --git a/dist/Carp/lib/Carp.pm b/dist/Carp/lib/Carp.pm
index 69caac3..6162177 100644
--- a/dist/Carp/lib/Carp.pm
+++ b/dist/Carp/lib/Carp.pm
@@ -24,7 +24,7 @@ BEGIN {
     }
 }
 
-our $VERSION = '1.29';
+our $VERSION = '1.30';
 
 our $MaxEvalLen = 0;
 our $Verbose    = 0;
@@ -431,10 +431,15 @@ sub trusts {
 sub trusts_directly {
     my $class = shift;
     no strict 'refs';
-    no warnings 'once';
-    return @{"$class\::CARP_NOT"}
-        ? @{"$class\::CARP_NOT"}
-        : @{"$class\::ISA"};
+    my $stash = \%{"$class\::"};
+    for my $var (qw/ CARP_NOT ISA /) {
+        # Don't try using the variable until we know it exists,
+        # to avoid polluting the caller's namespace.
+        if ( $stash->{$var} && @{"$class\::$var"} ) {
+           return @{"$class\::$var"}
+        }
+    }
+    return;
 }
 
 if(!defined($warnings::VERSION) ||
diff --git a/dist/Carp/t/Carp.t b/dist/Carp/t/Carp.t
index f7c226a..09c6652 100644
--- a/dist/Carp/t/Carp.t
+++ b/dist/Carp/t/Carp.t
@@ -3,7 +3,7 @@ no warnings "once";
 use Config;
 
 use IPC::Open3 1.0103 qw(open3);
-use Test::More tests => 61;
+use Test::More tests => 62;
 
 sub runperl {
     my(%args) = @_;
@@ -463,6 +463,16 @@ SKIP:
     );
 }
 
+{
+    package Foo::No::CARP_NOT;
+    eval { Carp::croak(1) };
+    ::is_deeply(
+        [ keys %Foo::No::CARP_NOT:: ],
+        [],
+        "Carp doesn't create CARP_NOT or ISA in the caller if they don't exist"
+    );
+}
+
 # New tests go here
 
 # line 1 "A"
-- 
1.7.12.4 (Apple Git-37)


```
</details>

**Comment by p5pRT** [comment]
### From @Hugmeir

<details><summary>0002-Carp-Don-t-autovivify-the-CARP_NOT-or-ISA-array.patch</summary>

```diff
From 0b08652ca5d81dfb384d12a0dedc3fb02c99fb43 Mon Sep 17 00:00:00 2001
From: Brian Fraser <fraserbn@gmail.com>
Date: Mon, 13 May 2013 16:37:22 -0300
Subject: [PATCH 2/2] Carp: Don't autovivify the CARP_NOT or ISA array

In this situation:

    use Carp;
    package Foo;
    $CARP_NOT = 1;
    warn *{$Foo::{CARP_NOT}}{ARRAY} || 'undefined';
    eval { Carp::croak(1) };
    warn *{$Foo::{CARP_NOT}}{ARRAY} || 'undefined';

The second warn shows that Carp is autovivifying the ARRAY slot
of the glob.  This commit builds on the previous and checks that
the array exists before trying to use it.
---
 dist/Carp/lib/Carp.pm |  4 ++--
 dist/Carp/t/Carp.t    | 10 +++++++++-
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/dist/Carp/lib/Carp.pm b/dist/Carp/lib/Carp.pm
index 6162177..bb557dd 100644
--- a/dist/Carp/lib/Carp.pm
+++ b/dist/Carp/lib/Carp.pm
@@ -435,8 +435,8 @@ sub trusts_directly {
     for my $var (qw/ CARP_NOT ISA /) {
         # Don't try using the variable until we know it exists,
         # to avoid polluting the caller's namespace.
-        if ( $stash->{$var} && @{"$class\::$var"} ) {
-           return @{"$class\::$var"}
+        if ( $stash->{$var} && *{$stash->{$var}}{ARRAY} && @{$stash->{$var}} ) {
+           return @{$stash->{$var}}
         }
     }
     return;
diff --git a/dist/Carp/t/Carp.t b/dist/Carp/t/Carp.t
index 09c6652..f51bd11 100644
--- a/dist/Carp/t/Carp.t
+++ b/dist/Carp/t/Carp.t
@@ -3,7 +3,7 @@ no warnings "once";
 use Config;
 
 use IPC::Open3 1.0103 qw(open3);
-use Test::More tests => 62;
+use Test::More tests => 63;
 
 sub runperl {
     my(%args) = @_;
@@ -471,6 +471,14 @@ SKIP:
         [],
         "Carp doesn't create CARP_NOT or ISA in the caller if they don't exist"
     );
+
+    package Foo::No::Autovivify;
+    $CARP_NOT = 1;
+    eval { Carp::croak(1) };
+    ::ok(
+        !defined *{$Foo::No::Autovivify::{CARP_NOT}}{ARRAY},
+        "Carp doesn't autovivify the CARP_NOT or ISA arrays if the globs exists but they lack the ARRAY slot"
+    );
 }
 
 # New tests go here
-- 
1.7.12.4 (Apple Git-37)


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon May 13 12&#8203;:58&#8203;:21 2013\, Hugmeir wrote&#8203;:
> This is a bug report for perl from fraserbn@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> In perl 5\.6\, this&#8203;:
> 
> perl \-MCarp \-le 'package Foo; eval \{ Carp&#8203;::croak\(1\) \}; print keys
> %Foo&#8203;::;'
> 
> Used to output 'ISA'\. In 5\.8 and newer\, it outputs "ISA" and
> "CARP\_NOT"\.
> This is because Carp&#8203;::trusts\_directly\(\) simply tries using both
> of those variables without checking if they exists first\, which
> ends up creating them\.
> The first of the attached patches fixes that so that no new globs are
> introduced to the caller's namespace\, by checking if the variable
> exists in the stash before using them\.
> 
> The second patch fixes a related issue&#8203;:
> 
>     use Carp;
>     package Foo;
>     $CARP\_NOT = 1;
>     warn \*\{$Foo&#8203;::\{CARP\_NOT\}\}\{ARRAY\} || 'undefined';
>     eval \{ Carp&#8203;::croak\(1\) \};
>     warn \*\{$Foo&#8203;::\{CARP\_NOT\}\}\{ARRAY\} || 'undefined';
> 
> Both should display "undefined"\, but instead the second warn
> shows that the array slot got vivified by Carp\.
> 

The patches look good\.  The code changes are plausible\.  Tests have been
added\.   The commit messages are descriptive\.

However\, when I applied the patch in a branch\, I got a test failure&#8203;:
\#\#\#\#\#
$ cd t;\./perl harness \-v \.\./dist/Carp/t/heavy\.t; cd \-
\.\./dist/Carp/t/heavy\.t \.\. 
1\.\.3
ok 1 \# control
ok 2 \# carp loaded by Carp&#8203;::Heavy
not ok 3 \# version numbers match
Failed 1/3 subtests 

Test Summary Report


---
\.\./dist/Carp/t/heavy\.t \(Wstat&#8203;: 0 Tests&#8203;: 3 Failed&#8203;: 1\)
&nbsp;  Failed test&#8203;:  3
Files=1\, Tests=3\,  0 wallclock secs \( 0\.00 usr \+  0\.01 sys =  0\.01 CPU\)
Result&#8203;: FAIL
\#\#\#\#\#
Now\, this is another case of a \(mindless\, IMO\) error in which a test is
comparing version numbers in two different files\.  What makes it
particularly peculiar in this instance is that there's nothing in
Carp&#8203;::Heavy any more \*except\* its version number\!

In any event\, I am attaching an additional patch which I believe
resolves this\.  Please review\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>117989--Bump-version-in-Carp-Heavy.patch</summary>

```diff
From 5b51dba777f23b14c0c9a9ae77462c273d6b5e65 Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Sun, 26 May 2013 04:52:00 +0200
Subject: [PATCH] Bump version in Carp::Heavy.

---
 dist/Carp/lib/Carp/Heavy.pm |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/dist/Carp/lib/Carp/Heavy.pm b/dist/Carp/lib/Carp/Heavy.pm
index 283b09b..db8b453 100644
--- a/dist/Carp/lib/Carp/Heavy.pm
+++ b/dist/Carp/lib/Carp/Heavy.pm
@@ -2,7 +2,7 @@ package Carp::Heavy;
 
 use Carp ();
 
-our $VERSION = '1.29';
+our $VERSION = '1.30';
 
 1;
 
-- 
1.7.1


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat May 25 20&#8203;:00&#8203;:14 2013\, jkeenan wrote&#8203;:
> On Mon May 13 12&#8203;:58&#8203;:21 2013\, Hugmeir wrote&#8203;:
> > This is a bug report for perl from fraserbn@&#8203;gmail\.com\,
> > generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> > 
> > In perl 5\.6\, this&#8203;:
> > 
> > perl \-MCarp \-le 'package Foo; eval \{ Carp&#8203;::croak\(1\) \}; print keys
> > %Foo&#8203;::;'
> > 
> > Used to output 'ISA'\. In 5\.8 and newer\, it outputs "ISA" and
> > "CARP\_NOT"\.
> > This is because Carp&#8203;::trusts\_directly\(\) simply tries using both
> > of those variables without checking if they exists first\, which
> > ends up creating them\.
> > The first of the attached patches fixes that

Thank you\.  Applied as 1a4f8f4\.

> > so that no new globs are
> > introduced to the caller's namespace\, by checking if the variable
> > exists in the stash before using them\.
> > 
> > The second patch fixes a related issue&#8203;:

Applied as 79f8d0e\.

> > 
> >     use Carp;
> >     package Foo;
> >     $CARP\_NOT = 1;
> >     warn \*\{$Foo&#8203;::\{CARP\_NOT\}\}\{ARRAY\} || 'undefined';
> >     eval \{ Carp&#8203;::croak\(1\) \};
> >     warn \*\{$Foo&#8203;::\{CARP\_NOT\}\}\{ARRAY\} || 'undefined';
> > 
> > Both should display "undefined"\, but instead the second warn
> > shows that the array slot got vivified by Carp\.
> > 
> 
> The patches look good\.  The code changes are plausible\.  Tests have been
> added\.   The commit messages are descriptive\.
> 
> However\, when I applied the patch in a branch\, I got a test failure&#8203;:
> \#\#\#\#\#
> $ cd t;\./perl harness \-v \.\./dist/Carp/t/heavy\.t; cd \-
> \.\./dist/Carp/t/heavy\.t \.\. 
> 1\.\.3
> ok 1 \# control
> ok 2 \# carp loaded by Carp&#8203;::Heavy
> not ok 3 \# version numbers match
> Failed 1/3 subtests 
> 
> Test Summary Report
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \.\./dist/Carp/t/heavy\.t \(Wstat&#8203;: 0 Tests&#8203;: 3 Failed&#8203;: 1\)
>   Failed test&#8203;:  3
> Files=1\, Tests=3\,  0 wallclock secs \( 0\.00 usr \+  0\.01 sys =  0\.01 CPU\)
> Result&#8203;: FAIL
> \#\#\#\#\#
> Now\, this is another case of a \(mindless\, IMO\) error in which a test is
> comparing version numbers in two different files\.  What makes it
> particularly peculiar in this instance is that there's nothing in
> Carp&#8203;::Heavy any more \*except\* its version number\!
> 
> In any event\, I am attaching an additional patch which I believe
> resolves this\.  Please review\.

Thank \*you\*\, too\.  Applied as 1ef9e4d\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
